### PR TITLE
fix(executor): Optional input artifacts. Fixes #2990

### LINKED ISSUE #2990: Cannot use `optional` with input artifacts/S3
Checklist:

 * [x] I've included the version. - master
 * [x] I've included reproduction steps.
 * [ ] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:

Started a step with an input artifact where the bucket does not exist, step failed.

**What you expected to happen**:

Artifact was marked optional - should pass.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:

- Argo version:
```
$ argo version
```
- Kubernetes version :
```
$ kubectl version -o yaml
```

**Other debugging information (if applicable)**:

- workflow result:

```
argo --loglevel DEBUG get <workflowname>
```

- executor logs:

```
kubectl logs <failedpodname> -c init
kubectl logs <failedpodname> -c wait
```

- workflow-controller logs:

```
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
```

**Logs**

```
argo get <workflowname>
kubectl logs <failedpodname> -c init
kubectl logs <failedpodname> -c wait
kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed the CLA.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My builds are green. Try syncing with master if they are not. 
* [x] My organization is added to USERS.md.

# Changes

* Each artifact repository that can output an artifact can now support optional input artifacts.
* HTTP also - as this was easy to test.

# Testing

| Name | Change | How |
|---|---|---|
| Artifactory | Translate 404 error | Ask @jessesuen or community |
| GCS | Bug fixes. Translate `ErrObjectNotExists`| Ask author @whynowy |
| Git | - | N/A |
| HDFS | Translate ErrNotExists | Ask @sarabala1979 and @dtaniwaki |
| HTTP | Translate exit code 22 | Automated |
| OSS | Bug fix | Ask @merlintang  |
| Raw | -| N/A |
| S3 | Translate error | Automated |

# Prototypes

https://bit.ly/argo-wf-protos

### CODE REVIEW

File: workflow/executor/executor.go [hunk 1]
```diff
@@ -164,6 +164,10 @@ func (we *WorkflowExecutor) LoadArtifacts() error {
 		tempArtPath := artPath + ".tmp"
 		err = artDriver.Load(&art, tempArtPath)
 		if err != nil {
+			if art.Optional && errors.IsCode(errors.CodeNotFound, err) {
```
Comment (simster7) [inline workflow/executor/executor.go:167]: Currently `Artifact.Optional` means "not specified". With this we also enhance it to mean "not found" as well. Is this what we want or should we distinguish between the two?

Some users might expect different behaviors with each of these.

Comment (alexec) [reply workflow/executor/executor.go:167]: Lets discus at stand-up tomorrow, I'm having trouble getting my head around how "optional" could not be interpreted as "optional" - where as something specified in the spec can be inferred to be "not specified". It's like words have new meanings.

Comment (alexec) [reply workflow/executor/executor.go:167]:
```
	// Make Artifacts optional, if Artifacts doesn't generate or exist
	Optional bool `json:"optional,omitempty" protobuf:"varint,8,opt,name=optional"`
```

Hmmm....

Comment (simster7) [reply workflow/executor/executor.go:167]: On further thought I agree that the two meanings are essentially the same.

"Optional" should be taken from the point of view of the container that receives the artifact. The reason as to why it's not there should not matter.

File: workflow/artifacts/hdfs/hdfs.go [hunk 2]
```diff
@@ -169,7 +169,14 @@ func (driver *ArtifactDriver) Load(inputArtifact *wfv1.Artifact, path string) er
 		}
 	}
 
-	return hdfscli.CopyToLocal(driver.Path, path)
+	err = hdfscli.CopyToLocal(driver.Path, path)
+	if err != nil {
```
Comment (dtaniwaki) [inline workflow/artifacts/hdfs/hdfs.go:173]: 👍

**Pull request merged**